File: src/main/java/Bf.java
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.PrimitiveIterator;
import java.util.function.Consumer;


final class Bf {

    interface Op extends Consumer<Tape> {
    }

    static final class LoopOp implements Op {
        final Op[] loop;

        LoopOp( Op[] loop ) {
            this.loop = loop;
        }

        @Override
        public void accept( Tape tape ) {
            while ( tape.get() > 0 ) for ( Op op : loop ) {
                op.accept( tape );
            }
        }
    }

    static final class PrintOp implements Op {
        @Override
        public void accept( Tape tape ) {
            System.out.print( ( char ) tape.get() );
        }
    }

    static final class Tape {
        private int[] tape;
        private int pos;

        Tape() {
            tape = new int[ 1 ];
        }

        int get() {
            return tape[ pos ];
        }

        void inc( int x ) {
            tape[ pos ] += x;
        }

        void move( int x ) {
            pos += x;
            while ( pos >= tape.length ) {
                int[] tape = new int[ this.tape.length * 2 ];
                System.arraycopy( this.tape, 0, tape, 0, this.tape.length );
                this.tape = tape;
            }

        }
    }

    static final class Program {
        static final Op INCR = tape -> tape.inc( 1 );
        static final Op DECR = tape -> tape.inc( -1 );
        static final Op MV_UP = tape -> tape.move( 1 );
        static final Op MV_DOWN = tape -> tape.move( -1 );
        static final Op PRINT = new PrintOp();

        private final Op[] ops;

        Program( String code ) {
            ops = parse( code.chars().iterator() );
        }

        private Op[] parse( PrimitiveIterator.OfInt it ) {
            List<Op> res = new ArrayList<>();
            while ( it.hasNext() ) {
                switch ( it.nextInt() ) {
                    case '+':
                        res.add( INCR );
                        break;
                    case '-':
                        res.add( DECR );
                        break;
                    case '>':
                        res.add( MV_UP );
                        break;
                    case '<':
                        res.add( MV_DOWN );
                        break;
                    case '.':
                        res.add( PRINT );
                        break;
                    case '[':
                        res.add( new LoopOp( parse( it ) ) );
                        break;
                    case ']':
                        return res.toArray( new Op[ 0 ] );
                }
            }
            return res.toArray( new Op[ 0 ] );
        }

        void run() {
            Tape tape = new Tape();
            for ( Op op : ops ) op.accept( tape );
        }

    }

    public static void main( String[] args ) throws IOException {
        byte[] code = Files.readAllBytes( Paths.get( args[ 0 ] ) );

        long startTime = System.currentTimeMillis();
        Program program = new Program( new String( code, StandardCharsets.US_ASCII ) );
        program.run();
        System.err.printf( "time: %.3fs\n", ( System.currentTimeMillis() - startTime ) / 1e3 );
    }
}

File: src/main/kotlin/bf.kt
@file:JvmName("BfKotlin")

import java.io.File
import java.io.IOException
import java.io.PrintStream

sealed class Op {
    abstract fun run(tape: KtTape)

    object Incr : Op() {
        override fun run(tape: KtTape) = tape.inc(1)
    }

    object Decr : Op() {
        override fun run(tape: KtTape) = tape.inc(-1)
    }

    object MoveUp : Op() {
        override fun run(tape: KtTape) = tape.move(1)
    }

    object MoveDown : Op() {
        override fun run(tape: KtTape) = tape.move(-1)
    }

    class Print(private val out: PrintStream) : Op() {
        override fun run(tape: KtTape) {
            out.print(tape.get().toChar())
        }
    }

    class Loop(private val ops: Array<Op>) : Op() {
        override fun run(tape: KtTape) {
            while (tape.get() > 0) {
                runProgram(ops, tape)
            }
        }
    }
}

class KtTape {
    private var tape: IntArray = IntArray(1)
    private var pos: Int = 0

    val state: IntArray get() = tape.copyOf()

    fun get(): Int {
        return tape[pos]
    }

    fun inc(x: Int) {
        tape[pos] += x
    }

    fun move(x: Int) {
        pos += x
        while (pos >= tape.size) {
            val tape = IntArray(this.tape.size * 2)
            System.arraycopy(this.tape, 0, tape, 0, this.tape.size)
            this.tape = tape
        }
    }
}

private fun runProgram(program: Array<Op>, tape: KtTape) {
    for (op in program) {
        op.run(tape)
    }
}

class KtProgram(code: String, out: PrintStream) {
    private val ops: Array<Op>
    private val printer = Op.Print(out)

    init {
        val it = code.iterator()
        ops = parse(it)
    }

    private fun parse(it: CharIterator): Array<Op> {
        val res = arrayListOf<Op>()
        while (it.hasNext()) {
            when (it.next()) {
                '+' -> res.add(Op.Incr)
                '-' -> res.add(Op.Decr)
                '>' -> res.add(Op.MoveUp)
                '<' -> res.add(Op.MoveDown)
                '.' -> res.add(printer)
                '[' -> res.add(Op.Loop(parse(it)))
                ']' -> return res.toTypedArray()
            }
        }
        return res.toTypedArray()
    }

    fun run(): KtTape {
        val tape = KtTape()
        runProgram(ops, tape)
        return tape
    }
}

fun runWithTiming(runnable: () -> Any?) {
    val startTime = System.currentTimeMillis()
    runnable()
    System.err.printf("time: %.3fs\n", (System.currentTimeMillis() - startTime) / 1e3)
}

@Throws(IOException::class)
fun main(args: Array<String>) {
    val code = File(args[0]).readText(Charsets.US_ASCII)

    val runs = if (args.size > 1) args[1].toInt() else 1

    repeat(runs) {
        runWithTiming(KtProgram(code, System.out)::run)
    }
}
